### PR TITLE
Reordering plan optimizers

### PR DESCRIPTION
**Pull request by Praveen2112**
Presto's planning time increases as the number of columns in the given table increases, so before applying other optimizers for a given query it would be better if we remove the unnecessary column.
We first apply `PruneUnreferencedOutputs` optimizer then we apply the other optimizer. We tested this order on tpch tables on query 5

**On normal TPCH table**

Benchmark                     (iterativeOptimizerEnabled)    (stage)  Mode  Cnt   Score   Error  Units
BenchmarkPlanner.planQueries                         true  optimized  avgt   20  39.222 ± 3.603  ms/op
BenchmarkPlanner.planQueries                         true    created  avgt   20   3.349 ± 0.505  ms/op
BenchmarkPlanner.planQueries                        false  optimized  avgt   20  36.918 ± 4.752  ms/op
BenchmarkPlanner.planQueries                        false    created  avgt   20   3.530 ± 0.276  ms/op

 **Number of columns in a TPCH table are increased by 100 times**

Benchmark                     (iterativeOptimizerEnabled)    (stage)  Mode  Cnt     Score     Error  Units
BenchmarkPlanner.planQueries                         true  optimized  avgt   20  1103.009 ±  53.372  ms/op
BenchmarkPlanner.planQueries                         true    created  avgt   20    25.000 ±   1.107  ms/op
BenchmarkPlanner.planQueries                        false  optimized  avgt   20  1127.392 ± 117.608  ms/op
BenchmarkPlanner.planQueries                        false    created  avgt   20    31.669 ±   9.585  ms/op

**Benchmark after we reordered the optimizers**

Benchmark                     (iterativeOptimizerEnabled)    (stage)  Mode  Cnt   Score    Error  Units
BenchmarkPlanner.planQueries                         true  optimized  avgt   20  71.519 ±  4.426  ms/op
BenchmarkPlanner.planQueries                         true    created  avgt   20  25.124 ±  0.826  ms/op
BenchmarkPlanner.planQueries                        false  optimized  avgt   20  99.174 ± 29.015  ms/op
BenchmarkPlanner.planQueries                        false    created  avgt   20  23.561 ±  0.655  ms/op

The planning time is reduced from **1103ms** to **71ms**

### CODE REVIEW

**Comment by Praveen2112** [comment]
@kokosing Can you please review this ?

**Comment by martint** [comment]
We should use the rule-based optimizers (see columnPruningRules above) unless there’s a very good reason for the legacy optimizers.

**Comment by martint** [comment]
A couple of questions/comments:
* what does "TPCH table is increased by 100 times" mean? The planning time should not be sensitive to the size of the table.
* we should try to understand what's so expensive in those cases and work on optimizing that code path. While this will help if the columns are not referenced, if a query does need them it will be slow to plan.

**Comment by martint** [comment]
> Ideally it should happen but the planning time is sensitive to size of the table.

If you’re referring to number of columns, sure. When I say size I mean number of rows in a table, which is what choosing a higher scale factor will do.

**Comment by Praveen2112** [comment]
I was referring to the number of columns

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java [hunk 1]
```diff
@@ -230,6 +230,8 @@ public PlanOptimizers(
         PlanOptimizer predicatePushDown = new StatsRecordingPlanOptimizer(optimizerStats, new PredicatePushDown(metadata, sqlParser));
 
         builder.add(
+                // Clean up all other symbols or expressions which are not a part of the given query
+                new PruneUnreferencedOutputs(),
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: can you use `columnPruningRules`

Comment (Praveen2112) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: Yes. I will replace it with `IterativeOptimizers` but some tests were failed when we rearrange the optimizers. So will identify that and update things.

**Pull request closed without merging**